mlrun/k8s_utils.py

```diff
@@ -228,3 +230,344 @@ def handle_invalid(message):
+    node_selector: dict[str, str],
+    tolerations: list[kubernetes.client.V1Toleration],
+    affinity: typing.Optional[kubernetes.client.V1Affinity],
+) -> tuple[
+    dict[str, str],
+    list[kubernetes.client.V1Toleration],
+    typing.Optional[kubernetes.client.V1Affinity],
+]:
+    """
+    Enriches a pod spec's scheduling configuration (node selector, tolerations, affinity)
+    based on the provided preemption mode.
+
+    If no preemptible node configuration is defined in the system, or the mode is `none`,
+    the original values are returned unchanged.
+
+    Modes:
+        - allow: Adds tolerations, removes preemption constraints.
+        - constrain: Requires preemptible node affinity and adds tolerations.
+        - prevent: Enforces scheduling on non-preemptible nodes using taints or anti-affinity.
+        - none: No enrichment is applied.
+    """
+    if (
+        not mlconfig.is_preemption_nodes_configured()
+        or preemption_mode == mlrun.common.schemas.PreemptionModes.none.value
+    ):
+        return node_selector, tolerations, affinity
+
+    if not preemption_mode:
+        preemption_mode = mlconfig.function_defaults.preemption_mode
+        mlrun.utils.logger.debug(
+            "No preemption mode provided, using default",
+            default_preemption_mode=preemption_mode,
+        )
+
+    enriched_node_selector = copy.deepcopy(node_selector or {})
+    enriched_tolerations = copy.deepcopy(tolerations or [])
+    enriched_affinity = copy.deepcopy(affinity)
+    preemptible_tolerations = generate_preemptible_tolerations()
+
+    if handler := _get_mode_handler(preemption_mode):
+        enriched_node_selector, enriched_tolerations, enriched_affinity = handler(
+            enriched_node_selector,
+            enriched_tolerations,
+            enriched_affinity,
+            preemptible_tolerations,
+        )
+
+    return (
+        enriched_node_selector,
+        enriched_tolerations,
+        _prune_empty_affinity(enriched_affinity),
+    )
+
+
+def _get_mode_handler(mode: str):
+    return {
+        mlrun.common.schemas.PreemptionModes.prevent: _handle_prevent_mode,
+        mlrun.common.schemas.PreemptionModes.constrain: _handle_constrain_mode,
+        mlrun.common.schemas.PreemptionModes.allow: _handle_allow_mode,
+    }.get(mode)
+
+
+def _handle_prevent_mode(
+    node_selector: dict[str, str],
+    tolerations: list[kubernetes.client.V1Toleration],
+    affinity: typing.Optional[kubernetes.client.V1Affinity],
+    preemptible_tolerations: list[kubernetes.client.V1Toleration],
+) -> tuple[
+    dict[str, str],
+    list[kubernetes.client.V1Toleration],
+    typing.Optional[kubernetes.client.V1Affinity],
+]:
+    def _prune_tolerations(to_remove: list[kubernetes.client.V1Toleration]):
+        return [t for t in tolerations if t not in to_remove]
+
+    # Ensure no preemptible node tolerations
+    tolerations = _prune_tolerations(preemptible_tolerations)
```

Comment:
I don't see where in this method you reuse `_prune_tolerations`, so why not just set the `tolerations` variable to that list comprehension?